### PR TITLE
Filling-out `GRPCWriteTest` functional test

### PR DESCRIPTION
**Pull request by willvedd**
## Description
As noted in #775 , the functional test for the GRPC write operation is empty (along with several others). This PR fills out the `GRPCWriteTest` test with several cases.

## References
- Related Issue: #775 
- PR introducing empty test: #83 

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: cmd/openfga/main_test.go [hunk 1]
```diff
@@ -569,7 +569,130 @@ func connect(t *testing.T, tester OpenFGATester) *grpc.ClientConn {
...
+				Writes: &openfgav1.WriteRequestWrites{
+					TupleKeys: []*openfgav1.TupleKey{
+						{Object: "document:1", Relation: "viewer", User: "user:jon"},
+						{Object: "document:1", Relation: "viewer", User: "user:*"},
+					},
+				},
+			},
+			output: output{
+				errorCode:    codes.Code(2000),
+				errorMessage: "the typed wildcard 'user:*' is not an allowed type restriction for 'document#viewer'",
```
Comment (miparnisari) [inline cmd/openfga/main_test.go:676]: This logic is already tested here: https://github.com/openfga/openfga/blob/65283c2e627814c6130cd8127ed0f629895afded/internal/validation/validation_test.go#L510

I think having another one here doesn't add any value

File: cmd/openfga/main_test.go [hunk 2]
```diff
@@ -569,7 +569,130 @@ func connect(t *testing.T, tester OpenFGATester) *grpc.ClientConn {
...
+				AuthorizationModelId: modelID,
+				Writes: &openfgav1.WriteRequestWrites{
+					TupleKeys: []*openfgav1.TupleKey{
+						{Object: "document:1", Relation: "viewer", User: "user:jon"},
+					},
+				},
+			},
+			output: output{
+				errorCode: codes.OK,
+			},
```
Comment (miparnisari) [inline cmd/openfga/main_test.go:624]: Can you add a similar one where `Writes` is empty but `Deletes` is not empty? It should also give OK

File: cmd/openfga/main_test.go [hunk 3]
```diff
@@ -569,7 +569,130 @@ func connect(t *testing.T, tester OpenFGATester) *grpc.ClientConn {
...
+				AuthorizationModelId: "invalid-model-id",
+				Writes:               &openfgav1.WriteRequestWrites{},
+			},
+			output: output{
+				errorCode:    codes.InvalidArgument,
+				errorMessage: `value does not match regex pattern "^[ABCDEFGHJKMNPQRSTVWXYZ0-9]{26}$`,
+			},
+		},
+		{
+			name: "empty-writes",
```
Comment (miparnisari) [inline cmd/openfga/main_test.go:651]:
```suggestion
			name: "empty-writes-and-deletes",
```

File: cmd/openfga/main_test.go [hunk 4]
```diff
@@ -569,7 +569,130 @@ func connect(t *testing.T, tester OpenFGATester) *grpc.ClientConn {
...
+			input: &openfgav1.WriteRequest{
+				StoreId:              storeID,
+				AuthorizationModelId: modelID,
+				Writes:               &openfgav1.WriteRequestWrites{},
+			},
+			output: output{
+				errorCode:    codes.InvalidArgument,
+				errorMessage: `invalid WriteRequestWrites.TupleKeys: value must contain at least 1 item(s)`,
```
Comment (miparnisari) [inline cmd/openfga/main_test.go:659]: Heh, I just noticed that this error message is ... sliiightly incorrect. It should say `either provide Writes or provide Deletes` ideally..

File: cmd/openfga/main_test.go [hunk 5]
```diff
@@ -569,7 +569,130 @@ func connect(t *testing.T, tester OpenFGATester) *grpc.ClientConn {
...
+	require.NotPanics(t, func() { ulid.MustParse(writeModelResp.GetAuthorizationModelId()) })
+	modelID := writeModelResp.GetAuthorizationModelId()
+
+	tests := []struct {
+		name   string
+		input  *openfgav1.WriteRequest
+		output output
+	}{
+		{
+			name: "happy-path",
```
Comment (jon-whit) [inline cmd/openfga/main_test.go:612]: Please use underscores as it is easier to find failed tests. We've defaulted to `_` where we can.

File: cmd/openfga/main_test.go [hunk 6]
```diff
@@ -569,7 +569,130 @@ func connect(t *testing.T, tester OpenFGATester) *grpc.ClientConn {
 }
 
 func GRPCWriteTest(t *testing.T, tester OpenFGATester) {
+	type output struct {
+		errorCode    codes.Code
+		errorMessage string
```
Comment (jon-whit) [inline cmd/openfga/main_test.go:574]: The error message really isn't part of the API contract. I'd encourage us to functionally test just the public API contract.

File: cmd/openfga/main_test.go [hunk 7]
```diff
@@ -569,7 +569,130 @@ func connect(t *testing.T, tester OpenFGATester) *grpc.ClientConn {
...
+				errorCode:    codes.InvalidArgument,
+				errorMessage: `value does not match regex pattern "^[ABCDEFGHJKMNPQRSTVWXYZ0-9]{26}$`,
+			},
+		},
+		{
+			name: "empty-writes",
+			input: &openfgav1.WriteRequest{
+				StoreId:              storeID,
+				AuthorizationModelId: modelID,
+				Writes:               &openfgav1.WriteRequestWrites{},
```
Comment (jon-whit) [inline cmd/openfga/main_test.go:655]: What about if `Writes` and `Deletes` is nil? Mind adding a test for that.

**Review by jon-whit — CHANGES_REQUESTED** [review]
Left a few minor pieces of feedback. Thanks for the contribution!

File: cmd/openfga/main_test.go (hunk 1 above)
Comment (willvedd) [reply cmd/openfga/main_test.go:676]: Certainly some redundancy between unit tests and functional test but perhaps the value is ensuring that the validation function is integrated?

File: cmd/openfga/main_test.go (hunk 4 above)
Comment (willvedd) [reply cmd/openfga/main_test.go:678]: So there seems to be a bit of nuance here. The validator alerts on the first validation error if a non-nil `Writes` or `Deletes` value is provided but if both are nil, the error is basically what you noted above. I made another case for this LMKWYT.

File: cmd/openfga/main_test.go (hunk 6 above)
Comment (willvedd) [reply cmd/openfga/main_test.go:575]: Understood that error messages aren't strictly part of the contract but our API should deliver readable, correct error messages, no? Changes in messages probably should be caught too. Definitely a gray area, but just want to ensure that users don't lose ability to resolve errors. 

Granted, I'm coming in very green so maybe the above aren't quite concerns. Happy to remove if you feel strongly.

File: cmd/openfga/main_test.go (hunk 1 above)
Comment (willvedd) [reply cmd/openfga/main_test.go:694]: **Update:** Was explained that there was ample testing elsewhere to ensure that the validation is hooked up. Am removing this case.

Comment (miparnisari) [reply cmd/openfga/main_test.go:691]: > Am removing this case.

Did you forget to commit? I still see it 🤔

Comment (willvedd) [reply cmd/openfga/main_test.go]: Yes, my bad 🙈. Here is the removal: https://github.com/openfga/openfga/pull/1242/commits/afc7479a21f5db89fa6ef89f62c6d52adbced818

**Review by miparnisari — APPROVED** [review]

File: cmd/openfga/main_test.go (hunk 6 above)
Comment (willvedd) [reply cmd/openfga/main_test.go:575]: Going to resolve this so we can get this merged in but still interested in hearing your takes on this and if necessary will remove with a follow-up.

**Review by jon-whit — APPROVED** [review]

**Pull request merged**